### ISSUE TITLE
[Question] Why collected actions in the buffer are not the clipped action (continuous action space case) ?

### ISSUE THREAD

**Issue by TheoZan** [DLR-RM/stable-baselines3#1431]
### ❓ Question

Hello all,

I'm a bit confused by some of the implementation logic of on-policy algorithm in continuous action space settings.

When we collect experience from the current policy and fill the rollout buffer, the action that we take at each step is indeed a clipped action to avoid the env crashing caused by an invalid action taking.

```
# Clip the actions to avoid out of bound error
            if isinstance(self.action_space, spaces.Box):
                clipped_actions = np.clip(actions, self.action_space.low, self.action_space.high)

            new_obs, rewards, dones, infos = env.step(clipped_actions)
```

However, the action that is saved in the buffer is the actual action sampled from the policy network.

`rollout_buffer.add(self._last_obs, actions, rewards, self._last_episode_starts, values, log_probs)`

I understand the fact that we want to learn from what the NN has outputed to be able to correct its action taking but this behavior associates a reward that correponds to the upper bound (in the case we took an action > 1) to an action that is > 1.
How come the agent can learn to take actions within the actual range without falling into a sub optimal solution of taking only extreme actions yeilding the reward coresponding to the bounds. 

Furthermore, how associating the same reward corresponding to the bound to all actions that are beyond can help the agent to take actions within the bounds ?

Does resampling or redrawing a distribution to get valid actions a stupid idea to help the agent focus on valid action rather than learning the bounds ?

Note: I'm talking especially about PPO where actions are not squashed (?) but rather clipped.

Thank you, hope this makes sense.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by TheoZan** [comment]
maybe a duplicate of  #1430, sorry for that

**Comment by araffin** [comment]
Related: https://github.com/hill-a/stable-baselines/issues/36#issuecomment-425732532 and https://github.com/hill-a/stable-baselines/pull/42
(if you don't use actions that were sampled from the distribution, you might end up with NaNs)

**Comment by TheoZan** [comment]
thanks for the references.

I'm wondering if there is a technique to mask a part of the action space in continuous settings as we have in discrete action space where you can mask the logits ? In the case of state dependant action constraints. Otherwise the agent has to learn the boundaries, the state dependant constraints as well as the optimal policy.

Could it be possible to use prior information concerning a valid state dependant action space to reduce it to only its valid part ? 
If possible, is it even profitable (concidering we know the boundaries) ?

I guess its a bit far from the original question but its intriguing.
